ckan/lib/helpers.py

```diff
@@ -346,193 +312,91 @@ def url_for(*args: Any, **kw: Any) -> str:
-
-    URLs built by Pylons use the Routes syntax::
-
-        url_for(controller='my_ctrl', action='my_action', id='my_dataset')
-        # Returns '/dataset/my_dataset'
-
-    Or, using a named route::
-
-        url_for('dataset.read', id='changed')
-        # Returns '/dataset/changed'
-
-    Use ``qualified=True`` for a fully qualified URL when targeting a Pylons
-    endpoint.
-
-    For backwards compatibility, an effort is made to support the Pylons syntax
-    when building a Flask URL, but this support might be dropped in the future,
-    so calls should be updated.
     '''
-    # Get the actual string code for the locale
-    locale = kw.pop('locale', None)
-    if locale and isinstance(locale, i18n.Locale):
-        locale = i18n.get_identifier_from_locale_class(locale)
-
-    # remove __ckan_no_root and add after to not pollute url
-    no_root = kw.pop('__ckan_no_root', False)
-
-    _auto_flask_context = _get_auto_flask_context()
-    try:
-        if _auto_flask_context:
-            _auto_flask_context.push()
-
-        # First try to build the URL with the Flask router
-        # Temporary mapping for pylons to flask route names
-        if len(args):
-            args = (map_pylons_to_flask_route_name(args[0]),)
-        my_url = _url_for_flask(*args, **kw)
-
-    except FlaskRouteBuildError:
-        raise
-    finally:
-        if _auto_flask_context:
-            _auto_flask_context.pop()
 
-    # Add back internal params
-    kw['__ckan_no_root'] = no_root
-
-    # Rewrite the URL to take the locale and root_path into account
-    return _local_url(my_url, locale=locale, **kw)
-
-
-def _url_for_flask(*args: Any, **kw: Any) -> str:
-    '''Build a URL using the Flask router
+    # these variables set by `nav_link` helper and they must be removed or else
+    # you'll see them in query string. It's better to modify `nav_link` so that
+    # it doesn't adds unnecesarry arguments and doesn't lock us out of using
+    # actual query parameters with these names.
+    blueprint = kw.pop("controller", None)
+    view = kw.pop("action", None)
+    if not args:
+        args = (f"{blueprint}.{view}",)
 
-    This function should not be called directly, use ``url_for`` instead
+    # pylons-style absolute URLs. It should be removed after explicit
+    # deprecation.
+    if kw.pop("qualified", False):
+        kw.setdefault("_external", True)
+    if not request:
+        kw.setdefault("_external", False)
 
-    This function tries to support the Pylons syntax for ``url_for`` and adapt
-    it to the Flask one, eg::
+    locale: str | i18n.Locale | None = kw.pop('locale', None)
 
-        # Pylons
-        url_for(controller='api', action='action', ver=3, qualified=True)
+    _safe_context = _get_auto_flask_context() or contextlib.nullcontext()
+    with _safe_context:
+        url = flask.url_for(*args, **kw)
 
-        # Flask
-        url_for('api.action', ver=3, _external=True)
+    if locale != 'default':
```

Comment:
locale will be None here if it wasn't included, shouldn't that be treated as the default?